### PR TITLE
REST and CLI API for domains

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR addresses "REST and CLI API for domains".

Proposed changes include:
* insert_domain
* enable/disable
* select_domain over REST.

### CODE REVIEW

File: src/domain/mongoose_domain_h.erl [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    {[{{<<"application">>, <<"json">>, '*'}, to_json}], Req, State}.
+
+to_json(Req, State) ->
+    ExtDomain = cowboy_req:binding(domain, Req),
+    Domain = jid:nameprep(ExtDomain),
+    case mongoose_domain_sql:select_domain(Domain) of
+        {ok, Props} ->
+            {jiffy:encode(Props), Req, State};
+        {error, not_found} ->
+            {stop, reply_error(404, <<"domain not found">>, Req), State}
```
Comment (NelsonVides) [inline src/domain/mongoose_domain_h.erl:41]: Why no test for a domain not found?

**Review by NelsonVides — COMMENTED** [review]
I'm afraid that `mongoose_domain_h` and `service_admin_extra_domain` are insanely similar to each other, I'm sad about the repeated code 😿 But I don't care about that, I'm mostly concerned, why basically not a single error-path has test coverage? Are we only testing happy cases? Can that be improved?

File: doc/developers-guide/domain_management.md [hunk 2]
```diff
@@ -118,10 +118,121 @@ The number of seconds after an event must be deleted from the `domain_events` ta
...
 # REST API
 
-We must provide REST API for Add/Remove and Enable/Disable interfaces of the
-MongooseIM service described in the section above.
+Provides for adding/removing and enabling/disabling domains over HTTP.
 
-# Command Line Interfaces
+Implemented by `mongoose_domain_handler` module.
+
+Confiugration example:
```
Comment (NelsonVides) [inline doc/developers-guide/domain_management.md:125]:
```suggestion
Configuration example:
```

File: doc/developers-guide/domain_management.md [hunk 3]
```diff
@@ -118,10 +118,121 @@ The number of seconds after an event must be deleted from the `domain_events` ta
...
+* 500 - other errors.
+
+Example of the result body with a failure reason:
+
+```
+{"what":"unknown host type"}
+```
+
+Check `src/domain/mongoose_domain_handler.erl` file for the the exact values of `what` field
+if needed.
```
Comment (NelsonVides) [inline doc/developers-guide/domain_management.md:162]:
```suggestion
Check the `src/domain/mongoose_domain_handler.erl` file for the exact values of the `what` field if needed.
```

File: doc/developers-guide/domain_management.md [hunk 4]
```diff
@@ -118,10 +118,121 @@ The number of seconds after an event must be deleted from the `domain_events` ta
...
+    -d '{"host_type": "type1"}'
+```
+
+Result codes:
+
+* 204 - the domain removed or not found.
+* 403 - the domain is static.
+* 403 - the DB service disabled.
+* 403 - the host type is wrong (does not match the host type in the database).
+* 403 - the unknown host type.
```
Comment (NelsonVides) [inline doc/developers-guide/domain_management.md:181]:
```suggestion
* 204 - the domain is removed or not found.
* 403 - the domain is static.
* 403 - the DB service is disabled.
* 403 - the host type is wrong (does not match the host type in the database).
* 403 - the host type is unknown.
```

**Review by NelsonVides — COMMENTED** [review]
Test coverage is much better now, but codecov shows some missing coverage for deleting domains errors, can we throw a bunch of tests there as well?

File: big_tests/tests/service_domain_db_SUITE.erl [hunk 5]
```diff
@@ -412,6 +434,116 @@ db_out_of_sync_crashes_node(_) ->
...
+    {error, not_found} = select_domain(mim(), <<"example.db">>).
+
+cli_cannot_delete_domain_without_correct_type(Config) ->
+    {"Added\n", 0} =
+        ejabberdctl("insert_domain", [<<"example.db">>, <<"type1">>], Config),
+    {"Error: \"wrong host type\"\n", 1} =
+        ejabberdctl("delete_domain", [<<"example.db">>, <<"type2">>], Config),
+    {ok, _} = select_domain(mim(), <<"example.db">>).
+
+cli_cannot_insert_domain_twice_with_the_another_host_type(Config) ->
```
Comment (chrzaszcz) [inline big_tests/tests/service_domain_db_SUITE.erl:468]: very minor: 'the another' looks awkward - it's a duplicate determiner, 'another' would suffice

File: big_tests/tests/service_domain_db_SUITE.erl [hunk 6]
```diff
@@ -412,6 +434,116 @@ db_out_of_sync_crashes_node(_) ->
...
+rest_delete_missing_domain(Config) ->
+    {{<<"204">>, _}, _} =
+        rest_delete_domain(<<"example.db">>, <<"type1">>).
+
+rest_cannot_enable_missing_domain(Config) ->
+    {{<<"404">>, <<"Not Found">>},
+     {[{<<"what">>, <<"domain not found">>}]}} =
+        rest_patch_enabled(<<"example.db">>, true).
+
+rest_cannot_insert_domain_twice_with_the_another_host_type(Config) ->
```
Comment (chrzaszcz) [inline big_tests/tests/service_domain_db_SUITE.erl:520]: 'the another' - see above

File: doc/developers-guide/domain_management.md [hunk 7]
```diff
@@ -118,10 +118,120 @@ The number of seconds after an event must be deleted from the `domain_events` ta
 
 # REST API
 
-We must provide REST API for Add/Remove and Enable/Disable interfaces of the
-MongooseIM service described in the section above.
+Provides for adding/removing and enabling/disabling domains over HTTP.
```
Comment (chrzaszcz) [inline doc/developers-guide/domain_management.md:121]:
```suggestion
Provides API for adding/removing and enabling/disabling domains over HTTP.
```

File: src/domain/service_admin_extra_domain.erl [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                           desc = "Disable a domain",
+                           module = ?MODULE, function = disable_domain,
+                           args = [{domain, binary}],
+                           result = {res, restuple}}
+    ].
+
+insert_domain(Domain, HostType) ->
+    SDomain = jid:nameprep(Domain),
+    case mongoose_domain_api:insert_domain(SDomain, HostType) of
+        ok -> {ok, "Added"};
```
Comment (chrzaszcz) [inline src/domain/service_admin_extra_domain.erl:41]: Minor: success string is capitalized, error string is not.

File: src/domain/service_admin_extra_domain.erl [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        {error, service_disabled} ->
+            {error, "service disabled"};
+        {error, unknown_host_type} ->
+            {error, "unknown host type"}
+    end.
+
+delete_domain(Domain, HostType) ->
+    SDomain = jid:nameprep(Domain),
+    case mongoose_domain_api:delete_domain(SDomain, HostType) of
+        ok -> {ok, "Deleted"};
```
Comment (chrzaszcz) [inline src/domain/service_admin_extra_domain.erl:55]: Minor: success string is capitalized, error string is not.

**Review by chrzaszcz — COMMENTED** [review]
Looks good. I added a few minor remarks. There are two more observations:
- Adding tests for unhappy paths in the REST API would be good. This can be done later.
- Newly added modules are missing specs, adding them in subsequent PR's would be nice as well.

File: src/domain/service_admin_extra_domain.erl (hunk 9 above)
Comment (arcusfelis) [reply src/domain/service_admin_extra_domain.erl:55]: Failed case prints:

```
Error: "unknown host type"
```

Success case prints:

```
Deleted
```

File: doc/developers-guide/domain_management.md [hunk 10]
```diff
@@ -118,10 +118,120 @@ The number of seconds after an event must be deleted from the `domain_events` ta
...
+```bash
+curl -v -X PUT "http://localhost:8088/api/domains/example.db" \
+    -H 'content-type: application/json' \
+    -d '{"host_type": "type1"}'
+```
+
+Result codes:
+
+* 204 - inserted.
+* 403 - domain already exists with a different host type.
```
Comment (chrzaszcz) [inline doc/developers-guide/domain_management.md:150]: I think `rest_cannot_insert_domain_twice_with_another_host_type` expects code 409 in this case.

**Review by chrzaszcz — COMMENTED** [review]
Good work! A lot of tests! Only one comment from me.

**Review by NelsonVides — APPROVED** [review]
All fine. Tests could have been organised to run in parallel when possible and then cut suite time, but anyway, we can leave that for the future.

**Pull request merged**